src/python/pants/engine/fs.py

```diff
@@ -212,6 +212,7 @@ class UrlToFetch:
 @dataclass(frozen=True)
 class Workspace:
   """Abstract handle for operations that touch the real local filesystem."""
+  side_effecting = True
```

Comment:
I'm curious, did you consider inheritance? 

I strongly prefer using a class field as you do here because multiple inheritance confuses me and inheritance opens up a can of worms of the superclass being able to add methods. But, I want to stress test this decision.

Reply:
cf. https://github.com/pantsbuild/pants/pull/8922#issuecomment-572327988 , I don't think marking this property via subclassing gets us anything marking it with a class variable doesn't

Reply:
Either an annotation (`@sideeffecting`) or a marker trait would be more self-documenting... someone seeing this property in isolation would not have anything to reference to figure out what it does, whereas with either of those alternatives they could go and inspect the docs on the type/annotation.

Another consideration would be: if we do decide to allow _types_ to be `uncacheable`, would we want to have a merged annotation to capture the fact that "sideeffecting, uncacheable, neither" is closer to being an enum? Or could that be handled with inheritance of the marker trait? etc.

Reply:
+1 to a class decorator. I agree with the benefits of the decorator being more formalized, such as being able to jump to the definition, whereas this field would require grepping for `side_effecting` to try to piece together what this does.

Ftr, the decorator could simply add the class property `side_effecting = True`. Your validation code would stay the same, only the call sites would be updated.

-1 on a trait mixin per my above comments on inheritance.